### PR TITLE
Blog

### LINKED ISSUE #122: Tiptap improvements
- Editor looks bad
- Missing critical functionality
- Hard to use

### PR DESCRIPTION
**Pull request by CoasterFan5**
Adds a blog to the website and updates text editing to be better in almost every aspect. 

Fixes #122

### CODE REVIEW

File: prisma/schema.prisma [hunk 1]
```diff
@@ -207,3 +208,42 @@ model OrgRole {
...
 	members      OrgUser[]
 }
+
+model BlogArticle {
+	createdAt     DateTime @default(now())
+	updatedAt     DateTime @updatedAt
+	id Int @id @default(autoincrement())
+	published Boolean @default(false)
+	articleURL String
+	articleImage String?
```
Comment (LeoDog896) [inline prisma/schema.prisma:218]: articleImage -> image

File: src/lib/articles.scss [hunk 2]
```diff
@@ -0,0 +1,37 @@
+.articleEdit,
```
Comment (LeoDog896) [inline src/lib/articles.scss:1]: using this as a stylesheet (`.scss`) will pollute the global stylesheet; are you sure you can't refactor this with componentization?

if not, I would instead inline this into a layout where you can [nest global styles as described in this SO post](https://stackoverflow.com/a/71329246/7589775).

File: src/lib/components/LavaLampBG.svelte [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+	let skipInitialTesting = false;
+	let disabledAnimation = false;
+
+	let canvasElement: HTMLCanvasElement;
+	let fakeCanvas: HTMLCanvasElement;
+
+	let context: CanvasRenderingContext2D | null;
+	let finalContext: CanvasRenderingContext2D | null;
+	$: if (canvasElement) {
```
Comment (LeoDog896) [inline src/lib/components/LavaLampBG.svelte:23]: you can remove this and silly boilerplate and height logic through [svelte-canvas](https://www.npmjs.com/package/svelte-canvas)

File: src/lib/utils/sanitizeTiptapContent.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import sanitizeHtml from 'sanitize-html';
+
+export const sanitizeTiptapContent = (content: string) => {
+	return sanitizeHtml(content, {
+		allowedClasses: {
+			'*': ['*']
```
Comment (LeoDog896) [inline src/lib/utils/sanitizeTiptapContent.ts:6]: this seems like [clickjacking](https://owasp.org/www-community/attacks/Clickjacking) just waiting to happen - no PoC I have in mind, though.

Comment (LeoDog896) [reply src/lib/utils/sanitizeTiptapContent.ts:6]: add a boolean or so, `trusted`, that allows custom classes and img for articles.

File: src/lib/utils/sanitizeTiptapContent.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+		},
+		allowedTags: [
+			'img',
```
Comment (LeoDog896) [inline src/lib/utils/sanitizeTiptapContent.ts:9]: you can IP-grab when images are arbitrary allowed without src blacklists.

File: src/routes/blog/(articles)/[article]/+page.server.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+			tagAssignments: {
+				include: {
+					tag: true
+				}
+			}
+		}
+	});
+
+	if (!article?.published) {
+		throw error(404, 'No article found');
```
Comment (LeoDog896) [inline src/routes/blog/(articles)/[article]/+page.server.ts:20]: on sveltekit 2 you don't have to throw error anymore

File: src/routes/blog/(articles)/[article]/+page.server.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+	}
+
+	if (!article) {
+		throw error(404, 'No article found');
```
Comment (LeoDog896) [inline src/routes/blog/(articles)/[article]/+page.server.ts:24]: on sveltekit 2 you don't have to throw error anymore

File: src/routes/blog/(articles)/[article]/+page.svelte [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+	import { Image } from '@unpic/svelte';
+	import dayjs from 'dayjs';
+	import hljs from 'highlight.js';
+	import { onMount } from 'svelte';
+
+	import Link from '$lib/components/Link.svelte';
+	import { sanitizeTiptapContent } from '$lib/utils/sanitizeTiptapContent.js';
+	onMount(() => {
+		hljs.highlightAll();
```
Comment (LeoDog896) [inline src/routes/blog/(articles)/[article]/+page.svelte:14]: https://www.npmjs.com/package/svelte-highlight

**Review by tristan-f-r — CHANGES_REQUESTED** [review]

File: src/routes/blog/(articles)/[article]/+page.svelte (hunk 8 above)
Comment (CoasterFan5) [reply src/routes/blog/(articles)/[article]/+page.svelte:14]: Im confused what benefits this has over just use highlight js

Comment (CoasterFan5) [reply src/routes/blog/(articles)/[article]/+page.svelte:14]: nvm

File: prisma/schema.prisma (hunk 1 above)
Comment (CoasterFan5) [reply prisma/schema.prisma:218]: not going to change <3

File: src/lib/utils/sanitizeTiptapContent.ts (hunk 4 above)
Comment (CoasterFan5) [reply src/lib/utils/sanitizeTiptapContent.ts]: resolved with 930cd3278af039c66085b0428697c497d7573b76

File: src/lib/components/LavaLampBG.svelte (hunk 3 above)
Comment (CoasterFan5) [reply src/lib/components/LavaLampBG.svelte:23]: I see no other advantages to this, and there is no point in refactoring working code using extra dependencies that work very similarly to the existing code IMO.

File: src/lib/articles.scss (hunk 2 above)
Comment (CoasterFan5) [reply src/lib/articles.scss:1]: Fixed in cace9c544b397d30ebfce925a835e0ff9d6cccd1

File: src/routes/api/image/upload/+server.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	let fileName = file.name;
+	if (fileName.length > 32) {
+		fileName = fileName.substring(fileName.length - 32);
+	}
+	const randomId = crypto.randomBytes(32).toString("base64url")
+	const key = `${user.id}/${randomId}/${fileName}`
+
+	await prisma.userImage.create({
```
Comment (LeoDog896) [inline src/routes/api/image/upload/+server.ts:45]: Move the `userImage` item creation to after the S3 object is created, or at least guarantee that the object won't exist if S3 fails.

File: src/lib/utils/sanitizeTiptapContent.ts [hunk 10]
```diff
@@ -1,5 +1,11 @@
 import sanitizeHtml from 'sanitize-html';
 
+const allowedHosts = [
+	"c.clubsaur.us",
+	"clubsaur.us",
+	"localhost"
```
Comment (LeoDog896) [inline src/lib/utils/sanitizeTiptapContent.ts]: only enable if on dev mode. just to be safe™️

**Review by tristan-f-r — CHANGES_REQUESTED** [review]

File: src/lib/utils/sanitizeTiptapContent.ts [hunk 11]
```diff
@@ -1,10 +1,6 @@
 import sanitizeHtml from 'sanitize-html';
 
-const allowedHosts = [
-	"c.clubsaur.us",
-	"clubsaur.us",
-	"localhost"
-]
+const allowedHosts = ['c.clubsaur.us', 'clubsaur.us', 'localhost'];
```
Comment (LeoDog896) [inline src/lib/utils/sanitizeTiptapContent.ts:3]: only enable localhost on dev mode

**Review by tristan-f-r — CHANGES_REQUESTED** [review]

File: src/lib/utils/sanitizeTiptapContent.ts (hunk 11 above)
Comment (CoasterFan5) [reply src/lib/utils/sanitizeTiptapContent.ts:3]: Fixed in [2d5e7ed](https://github.com/CoasterFan5/Clubsaurus/pull/150/commits/2d5e7edf1cb4c6c3eddd5a779cf62523a94d7443)

File: src/routes/api/image/upload/+server.ts (hunk 9 above)
Comment (CoasterFan5) [reply src/routes/api/image/upload/+server.ts]: Fixed in [2d5e7ed](https://github.com/CoasterFan5/Clubsaurus/pull/150/commits/2d5e7edf1cb4c6c3eddd5a779cf62523a94d7443)

**Review by tristan-f-r — APPROVED** [review]

**Pull request merged**